cpp/tests/experimental/louvain_primitives.cu

```diff
@@ -0,0 +1,179 @@
+//    however if they are not included prior to the two pattern
+//    includes below then the compilation fails.  I suspect these
+//    should be included somehow in the patterns includes below
+//
+#include <matrix_partition_device.cuh>
+#include <patterns/edge_op_utils.cuh>
+#include <utilities/host_scalar_comm.cuh>
+// END EXTRA INCLUDES
+
+#include <patterns/copy_v_transform_reduce_key_aggregated_out_nbr.cuh>
+#include <patterns/transform_reduce_by_adj_matrix_row_col_key_e.cuh>
+
+template <typename graph_t>
+void update_by_delta_modularity(raft::handle_t const& handle,
+                                graph_t const& graph_view,
+                                std::vector<typename graph_t::vertex_type> cluster_h,
+                                typename graph_t::weight_type total_edge_weight,
+                                typename graph_t::weight_type resolution)
+{
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t   = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+
+  vertex_t num_verts = graph_view.get_number_of_vertices();
+
+  rmm::device_uvector<vertex_t> cluster_v(num_verts, handle.get_stream());
+
+  raft::update_device(cluster_v.data(), cluster_h.data(), num_verts, handle.get_stream());
+
+  auto tmp = cugraph::experimental::transform_reduce_by_adj_matrix_col_key_e(
+    handle,
+    graph_view,
+    cluster_v.begin(),
+    cluster_v.begin(),
+    cluster_v.begin(),
+    [] __device__(auto r, auto c, auto w, auto rv, auto cv) {
+      printf("transform reduce (%d,%d,%g,%d,%d)\n", (int)r, (int)c, (float)w, (int)rv, (int)cv);
+      return w;
+    },
+    weight_t{0});
+
+  thrust::for_each(thrust::device,
+                   thrust::make_counting_iterator(0),
+                   thrust::make_counting_iterator(1),
+                   [b_key   = std::get<0>(tmp).begin(),
+                    e_key   = std::get<0>(tmp).end(),
+                    b_value = std::get<1>(tmp).begin(),
+                    e_value = std::get<1>(tmp).end()] __device__(auto) {
+                     char comma[2] = {0, 0};
+                     printf("keys = ");
+                     for (auto p = b_key; p != e_key; ++p) {
+                       printf("%s%d", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                     printf("values = ");
+                     comma[0] = (char)0;
+                     for (auto p = b_value; p != e_value; ++p) {
+                       printf("%s%g", comma, *p);
+                       comma[0] = ',';
+                     }
+                     printf("\n");
+                   });
+
+  if (graph_t::is_multi_gpu) {
+    cugraph::experimental::sort_and_shuffle_kv_pairs(
+      handle.get_comms(),
+      std::get<0>(tmp).begin(),
+      std::get<0>(tmp).end(),
+      std::get<1>(tmp).begin(),
+      [] __device__(auto) { return 0; },
+      handle.get_stream());
```

Comment:
```
+    communities = std::move(tmp_communities);
+    community_in_weight_sums = std::move(tmp_community_in_weight_sums);
```

Reply:
Actually, you can skip this and avoid defining tmp_... by using std::tie(communities, community_in_weight_sums) = cugraph::experimental::sort_and_shuffle_kv_pairs(...);

Also note that sort_and_shuffle_kv_pairs will be renamed to groupby_gpuids_and_shuffle_kv_pairs in https://github.com/rapidsai/cugraph/pull/1392

Reply:
Done